### ISSUE TITLE
Issues related to upgraded extension and recurring payments

### ISSUE THREAD

**Issue by laryn** [drastik/com.drastikbydesign.stripe#187]
I think/hope I've got myself sorted out after a lot of manual tinkering, but I just wanted to flag this here for people coming in the future. I have a site that has been using this extension for some time and has upgraded periodically, but always had some issues with recurring payments that I could never track down. I don't know if a fresh install will work out of the box with recurring payments now, but after I upgraded to the most recent 4.7-dev (as of a few days ago) to try to fix some of those issues I mentioned, I ran into the fact that Stripe was starting to receive errors on the webhook and left the webhook in a Pending state in the Stripe dashboard (probably due to [this one](https://github.com/drastik/com.drastikbydesign.stripe/issues/183)).

This, combined with the [Report Error Extension](https://github.com/mlutfy/ca.bidon.reporterror) helped me track down the issue(s) with this installation and fix them manually in the database as follows. Hopefully going forward things will work out of the box!

1. As mentioned, webhooks in Stripe were left in a pending state and retried periodically (guessing about once an hour). Every time they'd ping CiviCRM and generate an error (which was emailed to me courtesy of the [Report Error Extension](https://github.com/mlutfy/ca.bidon.reporterror).)
2. The error was: `Error relating this subscription id (sub_xyxyxyxyxxyxyxy) to the one in civicrm_stripe_subscriptions`
3. This led me to look in the `civicrm_stripe_subscriptions` table, and each subscription row was mostly empty of content. I went through to manually connect things as follows:

 - `customer_id` was filled in so I mapped this via the `civicrm_stripe_customers` table to a valid email address; I also mapped this to a valid CiviCRM Contact ID.
 - `subscription_id` was empty and I had to go through the Stripe `Subscriptions` page in the Stripe account itself, and copy out subscription_id's based on the email addresses pulled out above and manually fill this into the database
 - `contribution_recur_id` was also empty and I had to cross-reference and fill this information from the `civicrm_contribution_recur` table (via the CiviCRM Contact ID that I had written for each subscription) -- copying the `id` value from the `civicrm_contribution_recur` table.
 - `invoice_id` was also empty and I populated this with the `invoice_id` value from the same entry in the `civicrm_contribution_recur` table.

@drastik Do these mappings make sense? After I had done this the webhooks went from `Pending` to `Success` on the next attempt... and I'm guessing others may run into something similar if they've updated from older versions of the extension.

**Comment by h-c-c** [comment]
Hi @laryn,
When you upgraded, were you prompted to upgrade the database?  It sounds like that didn't happen. The Upgrader.php script should be doing all of this for you.

PS, invoice id is not relevant anymore.

**Comment by laryn** [comment]
@h-c-c Thanks for the response and for jogging my memory. Yes, I did get the prompt and try to do the upgrade. There were a few hiccups along the way which I happened to take a screenshot of:

<img width="638" alt="screen shot 2017-02-17 at 1 15 57 pm" src="https://cloud.githubusercontent.com/assets/871421/23179252/d69e49c4-f832-11e6-9558-b0417bfec7dc.png">

<img width="455" alt="screen shot 2017-02-17 at 1 16 20 pm" src="https://cloud.githubusercontent.com/assets/871421/23179259/dc09a0ca-f832-11e6-9439-e16aaa9657a3.png">

**Comment by h-c-c** [comment]
@laryn Thank you, I'm looking in this!  Quick question...which api version are you set to in Stripe?

**Comment by laryn** [comment]
<img width="453" alt="screen shot 2017-02-21 at 2 06 44 pm" src="https://cloud.githubusercontent.com/assets/871421/23182655/fbdacbe8-f83e-11e6-86e5-7abdb353cfba.png">

**Comment by h-c-c** [comment]
Ok, I think I see the issue.  The Upgrader.php script is currently set to quit if it encounters a customer that doesn't exist in Stripe anymore.   That could easily happen, since our webhook code never payed attention to those `customer.deleted` events if a customer was deleted in the Stripe UI.   

What's more likely, is there could be old subscriptions hanging around in there.  There was a time when we didn't do anything about `customer.subscription.deleted` either.  So this would cause an error as well.  

Basically, it should just soldier on, instead of quitting.   

This is really helpful....thanks so much @laryn!!
 
Patch on the way!
